tests/test_external_strategies.py

```diff
@@ -0,0 +1,119 @@
+"""
+Copyright 2018 Grid Singularity
+This file is part of D3A.
+
+This program is free software: you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation, either version 3 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+from unittest.mock import MagicMock, Mock
+
+from d3a.models.area import Area
+from d3a.models.strategy.external_strategies.load import LoadForecastExternalStrategy
+from d3a.models.strategy.external_strategies.pv import PVForecastExternalStrategy
+from d3a_interface.constants_limits import GlobalConfig
+from pendulum import now, Duration
+from pytest import mark
+
+
+def ext_strategy_fixture(strategy):
+    config = MagicMock()
+    config.ticks_per_slot = 90
+    GlobalConfig.end_date = GlobalConfig.start_date + Duration(days=1)
+    area = Area(name="forecast_pv", config=config, strategy=strategy,
+                external_connection_available=True)
+    parent = Area(name="parent_area", children=[area])
+    parent.activate()
+    area.activate()
+    strategy.connected = True
+    market = MagicMock()
+    market.time_slot = GlobalConfig.start_date
+    parent.get_future_market_from_id = lambda _: market
+    area.get_future_market_from_id = lambda _: market
+    return strategy
+
+
+class TestPVForecastExternalStrategy:
+
+    @mark.parametrize("strategy", [ext_strategy_fixture(LoadForecastExternalStrategy()),
+                                   ext_strategy_fixture(PVForecastExternalStrategy())])
+    def test_clear_energy_buffers(self, strategy):
```

Comment:
Nit: private methods are usually not considered for unit-tests (of course you can have exceptions). If you want more information about it, you can check this talk: https://youtu.be/URSWYvyc42M?t=947 . From this point, she talks about private methods and why not to test them.

Reply:
Thanks for sharing the link to the talk! 
I really can not connect it to this special case her though. Why wouldn't I test this unit (function) ?
I need to make sure that the buffers are cleared afterwards. Or how would you test this behaviour?
BTW; does the coverage calculation know about the rule "prefer not to test private methods"?

Reply:
Suppose you have a private method (`_clear_energy_buffers()` in this case). This private method will be called inside a public method (e.g. `event_market_cycle()`). We only want to test the _public_ side-effects and the outputs of our public method*.
Therefore, when testing `event_market_cycle()`, you will also check that `self.energy_forecast_buffer` and `self.energy_measurement_buffer` are empty. As you see, we are testing the public side-effects of `event_market_cycle()` and we don't care about which private methods are involved in the process (we only care about the final effect).

*Sidenote: ideally, a method should only do one thing: either return an output, or change the state of the object (not both things). It's called ["Command-Query separation" principle](https://en.wikipedia.org/wiki/Command%E2%80%93query_separation).

Reply:
Regarding the coverage calculation: the important thing is that your tests "hit" (execute) the code of the private methods. If you don't mock them, their code will be executed and should be therefore included in the coverage.